### PR TITLE
[IDSEQ-1578] Fix causes of potential duplicate collection location entries

### PR DESCRIPTION
**Pull request by jshoe**
### Description
- This fixes bugs that led to multiple entries occurring for the same locations, e.g. multiple "USAs".

### Notes
Two main bugs:
(1) We made a previous insufficiently-informed assumption that place_ids from the LocationIQ provider would be stable over time. This turned out not to be true, and we found some locations to change once a month or so. Ex: "USA" would get a new higher ID. So we have to fallback to string comparisons on, say, country/state/subdivision/city names.
(2) The location provider data has name changes that may not be internally consistent. For example, the entry for "Nevada" was updated to say its country is "United States of America" (instead of just "USA" before). But the entry for the country only has the name "USA". Therefore, we can resolve this by explicitly normalizing "United States of America" => "USA" for ourselves. Unfortunately we don't have real pointers to say "Nevada is in USA" besides the string fields, although we could also use the country_code in just this case.
- Docs: https://locationiq.com/docs-html/index.html#forward_response

Other fixes:
- Added some fixes for hardening against page crashes when I had inconsistent data in my local dev db.
- Catch invalid NaN coordinates.
- Fix for trying to get MetadataFields with no corresponding Metadatum entry.

### Tests
- See code tests.
- Debug test cases: Pick an entry you have in local db, e.g. "Nevada, USA". Update its locationiq_id to 0. Go through the upload flow and create a new sample with "Nevada, USA". See no duplicate entries for "Nevada" and "USA" were created.

### CODE REVIEW

File: app/helpers/location_helper.rb [hunk 1]
```diff
@@ -120,4 +120,18 @@ def self.filter_by_name(samples_with_metadata, query)
       samples.where("`metadata`.`string_validated_value` IN (?)", query)
     end
   end
+
+  # Normalize some location names from the provider for matching consistency.
```
Comment (gregdingle) [inline app/helpers/location_helper.rb]: it is surprising that we need to do this ourselves

File: config/initializers/location_name_aliases.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+begin
+  path = "app/lib/location_name_aliases.json"
+  LOCATION_NAME_ALIASES = JSON.parse(File.read(path))
+rescue SystemCallError, JSON::ParserError => err
+  Rails.logger.error("Couldn't load location name aliases. #{err.message}")
```
Comment (gregdingle) [inline config/initializers/location_name_aliases.rb]: wouldn't we want to crash here?

File: app/helpers/location_helper.rb [hunk 3]
```diff
@@ -21,12 +21,12 @@ def self.adapt_location_iq_response(body)
...
+      name: normalize_name_aliases(body["display_name"]),
       geo_level: geo_level,
-      country_name: address[country_key] || "",
-      state_name: address[state_key] || "",
-      subdivision_name: address[subdivision_key] || "",
-      city_name: address[city_key] || "",
+      country_name: normalize_name_aliases(address[country_key] || ""),
+      state_name: normalize_name_aliases(address[state_key] || ""),
+      subdivision_name: normalize_name_aliases(address[subdivision_key] || ""),
+      city_name: normalize_name_aliases(address[city_key] || ""),
```
Comment (gregdingle) [inline app/helpers/location_helper.rb]: did you consider running a migration to normalize the data that's already in prod?

File: app/helpers/samples_helper.rb [hunk 4]
```diff
@@ -587,9 +587,16 @@ def self.samples_by_metadata_field(sample_ids, field_name)
...
-        .includes(:metadata_field)
-        .group(Metadatum.where(key: field_name).first.validated_field)
+      # TODO(jsheu): Modify to get validated field from MetadataField.
+      first_datum = Metadatum.where(key: field_name).first
+      if first_datum
+        query
+          .includes(:metadata_field)
+          .group(first_datum.validated_field)
+      else
+        # No data for that field. Return empty grouped <ActiveRecord::Relation []> to be safe.
```
Comment (gregdingle) [inline app/helpers/samples_helper.rb]: 👍

File: app/helpers/location_helper.rb [hunk 5]
```diff
@@ -120,4 +120,18 @@ def self.filter_by_name(samples_with_metadata, query)
...
   end
+
+  # Normalize some location names from the provider for matching consistency.
+  # Ex: Treat "United States of America" and "USA" as the same in cases where
+  # the provider is not internally consistent.
+  #
+  # See config/initializers/location_name_aliases.rb and add important known
+  # aliases there.
+  def self.normalize_name_aliases(name)
+    if LOCATION_NAME_ALIASES && LOCATION_NAME_ALIASES.key?(name)
```
Comment (gregdingle) [inline app/helpers/location_helper.rb]: there is a chance that we could want two same-named entities to have different aliases. for example, washington the state and washington the city.

**Comment by jshoe** [comment]
> It's pretty crazy that's not true. Do they not provide _some_ id that is stable?
> And why do you think their location data is changing? If they are fixing bugs in their data we should try to benefit.

Unfortunately no. The core reason I've found is that LocationIQ (and many other similar providers) are heavily based on the OpenStreetMap project data, which can get 5 million changes a day. The OSM project has never had stable IDs and the same IDs can refer to very different locations over time. I suspect that LocationIQ does imports from OSM and other sources periodically and then just increments all their place_ids.

Related links:
- https://wiki.openstreetmap.org/wiki/Permanent_ID
- https://blog.emacsen.net/blog/2018/02/16/osm-is-in-trouble/ "Lack of Permanent IDs" section
- https://engineering.fb.com/ml-applications/mars/ OSM change management background

File: test/models/location_test.rb [hunk 6]
```diff
@@ -96,12 +96,13 @@ class LocationTest < ActiveSupport::TestCase
     assert mock.verify
   end
 
-  test "should find existing Location by API ID" do
+  test "should find existing Location by matching fields" do
+    location_info = { name: "Nevada, USA", geo_level: "state", country_name: "USA", state_name: "Nevada", subdivision_name: "", city_name: "" }
```
Comment (tfrcarvalho) [inline test/models/location_test.rb]: Should we write a test for not yet normalized fields?

File: app/helpers/location_helper.rb (hunk 3 above)
Comment (jshoe) [reply app/helpers/location_helper.rb]: I've corrected the existing ones already but I do plan to do it again manually 👍

File: config/initializers/location_name_aliases.rb (hunk 2 above)
Comment (jshoe) [reply config/initializers/location_name_aliases.rb]: Sure, changed it to re-raise and crash

File: app/helpers/location_helper.rb (hunk 5 above)
Comment (jshoe) [reply app/helpers/location_helper.rb]: That's a great point. Adapted to split by geo_levels 👍

File: app/models/location.rb [hunk 7]
```diff
@@ -86,27 +86,39 @@ def self.geosearch_by_osm_id(osm_id, osm_type)
...
     else
       # If osm_id and osm_type are missing, just use the original params.
       # 'New' without saving so make sure caller saves.
-      new_from_params(loc)
+      new_from_params(loc_info)
     end
   end
 
+  # Consider it a match if all fields to refer to the same place.
+  def self.find_with_fields(loc_info)
```
[the remark this answers is not in the document]
Comment (jshoe) [reply app/models/location.rb]: We had:
`t.index ["country_name", "state_name", "subdivision_name", "city_name"]`
`t.index ["geo_level"]`
`t.index ["name"]`

Theoretically MySQL could use an index merge with them, but I just added another composite with these 6 to optimize.

File: app/helpers/samples_helper.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (jshoe) [reply app/helpers/samples_helper.rb]: 👍

File: test/models/location_test.rb (hunk 6 above)
Comment (jshoe) [reply test/models/location_test.rb]: Discussed offline: the existing fields would still need manual or migration correction unless we also wanted to add the normalization to `find_with_fields`

**Pull request merged**